guest/file_manager.go

```diff
@@ -122,25 +122,26 @@ func (m FileManager) DeleteFile(ctx context.Context, auth types.BaseGuestAuthent
 // The InitiateFileTransfer{From,To}Guest methods return a URL with the host set to "*" when connected directly to ESX,
 // but return the address of VM's runtime host when connected to vCenter.
 func (m FileManager) TransferURL(ctx context.Context, u string) (*url.URL, error) {
-	url, err := m.c.ParseURL(u)
+	turl, err := url.Parse(u)
 	if err != nil {
 		return nil, err
 	}
 
+	needsHostname := turl.Hostname() == "*"
 	if !m.c.IsVC() {
-		return url, nil // we already connected to the ESX host and have its thumbprint
```

Comment:
Since the call to `m.c.ParseURL` is removed, we need to do this before returning: https://github.com/vmware/govmomi/blob/4316838a111c783776b1d9e06a6f7f61e507feeb/vim25/soap/client.go#L602

This should work with your patch:
```go
        if needsHostname {
                turl.Host = m.c.URL().Host
        }
```